Mathlib/RingTheory/MvPowerSeries/PiTopology.lean

```diff
@@ -0,0 +1,292 @@
+
+- `tendsto_pow_of_constantCoeff_nilpotent_iff` : the powers of `f` converge to zero iff
+the constant coefficient of `f` is nilpotent
+
+- `hasSum_of_monomials_self` : viewed as an infinite sum, a power series coverges to itself
+
+TODO: add the similar result for the series of homogeneous components
+
+## Instances
+
+- If `R` is a topological (semi)ring, then so is `MvPowerSeries σ R`
+
+- If the topology of `R` is T2, then so is that of `MvPowerSeries σ R`
+
+- If `R` is a `UniformAddGroup`, then so is `MvPowerSeries σ R``
+
+- If `R` is complete, then so is `MvPowerSeries σ R`
+
+-/
+
+
+theorem MvPowerSeries.apply_eq_coeff {σ R : Type _} [Semiring R] (f : MvPowerSeries σ R)
+    (d : σ →₀ ℕ) : f d = MvPowerSeries.coeff R d f :=
+  rfl
+
+namespace MvPowerSeries
+
+open Function
+
+variable {σ R : Type*}
+
+namespace WithPiTopology
+
+variable [TopologicalSpace R]
+
+variable (R) in
+/-- The pointwise topology on MvPowerSeries -/
+scoped instance : TopologicalSpace (MvPowerSeries σ R) :=
+  Pi.topologicalSpace
+
+variable [Semiring R]
+
+variable (R) in
+/-- coeff are continuous -/
+theorem continuous_coeff (d : σ →₀ ℕ) : Continuous (MvPowerSeries.coeff R d) :=
+  continuous_pi_iff.mp continuous_id d
+
+variable (R) in
+/-- constant_coeff is continuous -/
+theorem continuous_constantCoeff : Continuous (constantCoeff σ R) :=
+  continuous_coeff R 0
+
+/-- A family of power series converges iff it converges coefficientwise -/
+theorem tendsto_iff_coeff_tendsto {ι : Type*}
+    (f : ι → MvPowerSeries σ R) (u : Filter ι) (g : MvPowerSeries σ R) :
+    Filter.Tendsto f u (nhds g) ↔
+    ∀ d : σ →₀ ℕ, Filter.Tendsto (fun i => coeff R d (f i)) u (nhds (coeff R d g)) := by
+  rw [nhds_pi, Filter.tendsto_pi]
+  exact forall_congr' (fun d => Iff.rfl)
+
+variable (σ R)
+
+/-- The semiring topology on MvPowerSeries of a topological semiring -/
+@[scoped instance]
+theorem instTopologicalSemiring [TopologicalSemiring R] :
+    TopologicalSemiring (MvPowerSeries σ R) where
+    continuous_add := continuous_pi fun d => Continuous.comp continuous_add
+      (Continuous.prod_mk (Continuous.fst' (continuous_coeff R d))
+        (Continuous.snd' (continuous_coeff R d)))
+    continuous_mul := continuous_pi fun _ => continuous_finset_sum _ (fun i _ => Continuous.comp
+      continuous_mul (Continuous.prod_mk (Continuous.fst' (continuous_coeff R i.fst))
+        (Continuous.snd' (continuous_coeff R i.snd))))
+
+/-- The ring topology on MvPowerSeries of a topological ring -/
+@[scoped instance]
+theorem instTopologicalRing (R : Type*) [TopologicalSpace R] [Ring R] [TopologicalRing R] :
+    TopologicalRing (MvPowerSeries σ R) :=
+  { instTopologicalSemiring σ R with
+    continuous_neg := continuous_pi fun d ↦ Continuous.comp continuous_neg
+      (continuous_coeff R d) }
+
+/-- MvPowerSeries on a T2Space form a T2Space -/
+@[scoped instance]
+theorem instT2Space [T2Space R] : T2Space (MvPowerSeries σ R) where
+  t2 x y h := by
```

Comment:
Please reuse instances about general Pi types as much as possible. If we don't have some instance, then we should add it instead of proving it in a special case.

Reply:
We do. Thanks for noting.